packages/contractkit/src/utils/fp-ts.ts

```
@@ -0,0 +1,56 @@
+import fetch from 'cross-fetch'
+import { Either, isRight, left, right } from 'fp-ts/lib/Either'
+import { TaskEither, tryCatch } from 'fp-ts/lib/TaskEither'
+import { writeFileSync } from 'fs'
+
+export class NetworkError extends Error {}
+
+export function fetchAsTaskEither(url: string) {
+  return tryCatch(
+    () =>
+      fetch(url).then((x) => {
+        if (x.ok) {
+          return x.text()
+        } else {
+          throw new Error(`Request failed with status ${x.status}`)
+        }
+      }),
+    (reason: any) => new NetworkError(reason.toString())
```

not sure that i agree with this.... we are masking the real error here; which for the case of fetch error has more information that just the `toSTring()`

---

maybe do `reason => reason`

---

I wanted a way to have effectively a union type of the errors for the pipeline which is why I need to ensure all the errors. Maybe that's the wrong goal though